### ISSUE TITLE
Directory permissions: "The /flarum/app directory is not writable."

### ISSUE THREAD

**Issue by Simon-L** [mondediefr/docker-flarum#38]
As of the latest version, I get the same permission errors as reported in #36 
The full error I get is:
```
flarum            | [INFO] First launch, installation...
flarum            | Please fix the following errors before we can continue with the installation.
flarum            | The /flarum/app directory is not writable.
flarum            | Please chmod this directory and its contents to 0775.
flarum            | [INFO] End of flarum installation
```
During the many tests I've done, it also complains about the `/flarum/app/public/assets` directory.
I simplified the `docker-compose.yml` file to the following:
```
version: '3'

services:
    flarum:
        image: flarum-git
        container_name: flarum
        environment:
            - FORUM_URL=http://localhost:8888
            - DB_PASS=xxxxxx
            - DEBUG=true
            - FLARUM_ADMIN_USER=admin
            - FLARUM_ADMIN_PASS=xxxxxxx
            - FLARUM_ADMIN_MAIL=xxxxxxxxxx
        volumes:
            - /flarum/app/public/assets
            - /flarum/app/extensions
        depends_on:
            - flarum_mariadb

    flarum_mariadb:
        image: mariadb:10.2
        container_name: flarum_mariadb
        volumes:
            - /var/lib/mysql
        environment:
            - MYSQL_ROOT_PASSWORD=xxxxxxxxxx
            - MYSQL_DATABASE=flarum
            - MYSQL_USER=flarum
            - MYSQL_PASSWORD=xxxxxxxx
```

I have noted that if I the stop it right after that error, remove the `installed.txt` file and re-run `docker-compose up`, it does work in some combinations of permissions on the volumes that I can not even remember exactly given that I have tried so many that I can't recall!

I also tried setting the `GID` and `ID` variables to my  host user's value and chowning the volumes accordingly, to no avail.

Can anyone try and see if it's possible to reproduce?
Pull this image, copy/paste the file above, change the image name to the one you pull, edit the environnment variables, run it with `docker-compose up`, without -d to keep the logs foreground!

**Comment by hakimelek** [comment]
I am hitting the same issue ^

**Comment by Simon-L** [comment]
Good to know!
Made some more tests:
It seems the issue comes from the Dockerfile, `COPY rootfs /` runs *after* creation and chown of the `/flarum/app` folder.
COPY creates folders and files with ID and GID 0, writing over the files and folders that need to be writable.

One easy fix is __adding__ this line in Dockerfile after the first "COPY":
`COPY --chown=991:991 rootfs/flarum /flarum`
This technically copies the content of this folder twice, except the second time has the right permissions and ownership.

**Comment by Magicalex** [comment]
I watch this issue as soon as possible

**Comment by GTANAdam** [comment]
What the f#ck?

**Comment by Hydrog3n** [comment]
> What the f#ck? still watching as soon as possible?
> A one liner seems to hard to fix, check this guy out:
> https://github.com/jackiehope/docker-flarum
> He had also pushed it to docker
> https://hub.docker.com/r/jackieqiu/flarum
> y'all are so lazy

You can make a PR if you make the fix. 
Thank you to your help.

**Comment by hardware** [comment]
@Hydrog3n I banned this jerk. Do not thank him for insulting you, omg.

The real fix isn't that simple, @Simon-L proposed a temporary solution but I can't obviously push that in production.

I don't have time to fix this for the moment, I will try to look that later.

**Comment by hardware** [comment]
For now, I tag this repository as "Unmaintened" because no one seems to be able to maintain the docker image. For the moment, I don't have time, and we can not propose an obsolete image.

**Comment by Magicalex** [comment]
@Simon-L and @hakimelek have you still the same problem with flarum 0.1.0-beta.10 ?

**Comment by TrevorKS** [comment]
> environment:
>   - UID=nginx
>   - GID=nginx

I could not get this to work as a temp solution.

I am still experiencing this issue.

**Comment by TrevorKS** [comment]
Does anyone know of a temporary solution to this problem?

**Comment by Hydrog3n** [comment]
> Does anyone know of a temporary solution to this problem?
Hello, maybe with chown with user ID and group `chown 991:991 <volume_on_host>` the dir on your server.

**Comment by TrevorKS** [comment]
I seem to still be unable to fix the issue with the solutions in this thread.

**Comment by TrevorKS** [comment]
Sorry keep posting, but has anyone figured out a temporary solution to get this to run in docker? I couldnt get the tips mentioned in this thread to work. Anyone else have any working ideas? Is @Simon-L soultion still working? It seemed to fail for me

**Comment by Magicalex** [comment]
Can you give more information? Your docker-compose, os, cpu etc

**Comment by TrevorKS** [comment]
`docker-compose.yml` File

```
version: "3"

services:
  flarum:
    image: mondedie/docker-flarum:0.1.0-beta.10-stable
    container_name: flarum
    env_file:
      - ./flarum.env
    volumes:
      - ./Data/assets:/flarum/app/public/assets
      - ./Data/extensions:/flarum/app/extensions
      - ./Data/nginx:/etc/nginx/conf.d
    depends_on:
      - mariadb

  mariadb:
    image: mariadb:10.4
    container_name: mariadb
    environment:
      - MYSQL_ROOT_PASSWORD=testing123!
      - MYSQL_DATABASE=flarum
      - MYSQL_USER=flarum
      - MYSQL_PASSWORD=testing123!
    volumes:
      - /mnt/docker/mysql/db:/var/lib/mysql

networks:
  default:
    external:
      name: reverseproxy_default
```


`flarum.env` File
```
DEBUG=true
FORUM_URL=http://forums.motorboards.tech

# Database configuration
DB_HOST=mariadb
DB_NAME=flarum
DB_USER=flarum
DB_PASS=testing
DB_PREF=flarum_
DB_PORT=3306


# User admin flarum (environment variable for first installation)
# /!\ admin password must contain at least 8 characters /!\
FLARUM_ADMIN_USER=admin
FLARUM_ADMIN_PASS=testing123!
FLARUM_ADMIN_MAIL=TrevorKSmith@Motorboards.tech
FLARUM_TITLE=Test flarum
```


Below is my output of `docker-compose up`

```
root@ip-172-31-8-177:/srv/Docker/Flarum# docker-compose up
Creating mariadb ... done
Creating flarum  ... done
Attaching to mariadb, flarum
mariadb    | 2019-10-19 19:09:42 0 [Note] mysqld (mysqld 10.4.8-MariaDB-1:10.4.8+maria~bionic) starting as process 1 ...
mariadb    | 2019-10-19 19:09:42 0 [Note] InnoDB: Using Linux native AIO
mariadb    | 2019-10-19 19:09:42 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
mariadb    | 2019-10-19 19:09:42 0 [Note] InnoDB: Uses event mutexes
mariadb    | 2019-10-19 19:09:42 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
mariadb    | 2019-10-19 19:09:42 0 [Note] InnoDB: Number of pools: 1
mariadb    | 2019-10-19 19:09:42 0 [Note] InnoDB: Using SSE2 crc32 instructions
mariadb    | 2019-10-19 19:09:42 0 [Note] mysqld: O_TMPFILE is not supported on /tmp (disabling future attempts)
mariadb    | 2019-10-19 19:09:42 0 [Note] InnoDB: Initializing buffer pool, total size = 256M, instances = 1, chunk size = 128M
mariadb    | 2019-10-19 19:09:42 0 [Note] InnoDB: Completed initialization of buffer pool
mariadb    | 2019-10-19 19:09:42 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
flarum     | [INFO] First launch, installation...
flarum     | Please fix the following problems before we can continue with the installation.
flarum     | The /flarum/app directory is not writable.
flarum     | Please chmod this directory and its contents to 0775.
flarum     | The /flarum/app/public/assets directory is not writable.
flarum     | Please chmod this directory and its contents to 0775.
flarum     | [INFO] End of flarum installation
mariadb    | 2019-10-19 19:09:43 0 [Note] InnoDB: 128 out of 128 rollback segments are active.
mariadb    | 2019-10-19 19:09:43 0 [Note] InnoDB: Creating shared tablespace for temporary tables
mariadb    | 2019-10-19 19:09:43 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
mariadb    | 2019-10-19 19:09:43 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
mariadb    | 2019-10-19 19:09:43 0 [Note] InnoDB: Waiting for purge to start
mariadb    | 2019-10-19 19:09:43 0 [Note] InnoDB: 10.4.8 started; log sequence number 20274249; transaction id 14608
mariadb    | 2019-10-19 19:09:43 0 [Note] Plugin 'FEEDBACK' is disabled.
mariadb    | 2019-10-19 19:09:43 0 [Note] Server socket created on IP: '::'.
mariadb    | 2019-10-19 19:09:43 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
mariadb    | 2019-10-19 19:09:43 0 [Warning] 'proxies_priv' entry '@% root@8fcb80558ea1' ignored in --skip-name-resolve mode.
mariadb    | 2019-10-19 19:09:43 0 [Note] Reading of all Master_info entries succeeded
mariadb    | 2019-10-19 19:09:43 0 [Note] Added new Master_info '' to hash table
mariadb    | 2019-10-19 19:09:43 0 [Note] mysqld: ready for connections.
mariadb    | Version: '10.4.8-MariaDB-1:10.4.8+maria~bionic'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  mariadb.org binary distribution
mariadb    | 2019-10-19 19:09:43 0 [Note] InnoDB: Buffer pool(s) load completed at 191019 19:09:43

```

**Comment by Magicalex** [comment]
Can you try my fix with the image latest?

```
version: "3"

services:
  flarum:
    image: mondedie/docker-flarum:latest
```

**Comment by TrevorKS** [comment]
The original issue i believe has been fixed. It is no longer reporting permission issues.

A new issue is now occurring.

`[Warning] Access denied for user 'flarum'@'172.24.0.5' (using password: YES)
`

I remember this issue months ago when i attempted to use this image in the past. If i remember correctly, the solution was to change `localhost` to `mariadb` in the `config.php` file which is suppose to be found at the root of the flarum installation. I could not find this file this time.

`ping mariadb` from inside the flarum container successfully reports the database at `172.24.0.4` , which strongly makes me believe this is not a DNS issue.

For some reason, the flarum is still looking for the database on localhost, and not the remote host.

output

```
mariadb    | 2019-10-20 22:18:59 0 [Note] mysqld (mysqld 10.4.8-MariaDB-1:10.4.8+maria~bionic) starting as process 1 ...
flarum     | [INFO] First launch, installation...
mariadb    | 2019-10-20 22:18:59 0 [Note] InnoDB: Using Linux native AIO
mariadb    | 2019-10-20 22:18:59 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
mariadb    | 2019-10-20 22:18:59 0 [Note] InnoDB: Uses event mutexes
mariadb    | 2019-10-20 22:18:59 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
mariadb    | 2019-10-20 22:18:59 0 [Note] InnoDB: Number of pools: 1
mariadb    | 2019-10-20 22:18:59 0 [Note] InnoDB: Using SSE2 crc32 instructions
mariadb    | 2019-10-20 22:18:59 0 [Note] mysqld: O_TMPFILE is not supported on /tmp (disabling future attempts)
mariadb    | 2019-10-20 22:18:59 0 [Note] InnoDB: Initializing buffer pool, total size = 256M, instances = 1, chunk size = 128M
mariadb    | 2019-10-20 22:18:59 0 [Note] InnoDB: Completed initialization of buffer pool
mariadb    | 2019-10-20 22:18:59 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
mariadb    | 2019-10-20 22:19:00 0 [Note] InnoDB: 128 out of 128 rollback segments are active.
mariadb    | 2019-10-20 22:19:00 0 [Note] InnoDB: Creating shared tablespace for temporary tables
mariadb    | 2019-10-20 22:19:00 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
mariadb    | 2019-10-20 22:19:00 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
mariadb    | 2019-10-20 22:19:00 0 [Note] InnoDB: 10.4.8 started; log sequence number 23660721; transaction id 17567
mariadb    | 2019-10-20 22:19:00 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
mariadb    | 2019-10-20 22:19:00 0 [Note] Plugin 'FEEDBACK' is disabled.
mariadb    | 2019-10-20 22:19:00 0 [Note] Server socket created on IP: '::'.
mariadb    | 2019-10-20 22:19:00 0 [Warning] 'proxies_priv' entry '@% root@8fcb80558ea1' ignored in --skip-name-resolve mode.
mariadb    | 2019-10-20 22:19:00 0 [Note] Reading of all Master_info entries succeeded
mariadb    | 2019-10-20 22:19:00 0 [Note] Added new Master_info '' to hash table
mariadb    | 2019-10-20 22:19:00 0 [Note] mysqld: ready for connections.
mariadb    | Version: '10.4.8-MariaDB-1:10.4.8+maria~bionic'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  mariadb.org binary distribution
mariadb    | 2019-10-20 22:19:00 0 [Note] InnoDB: Buffer pool(s) load completed at 191020 22:19:00
flarum     | Installing Flarum...
mariadb    | 2019-10-20 22:19:12 8 [Warning] Access denied for user 'flarum'@'172.24.0.5' (using password: YES)
flarum     | Connecting to database...failed
flarum     | Rolling back...
flarum     | 
flarum     | In Pipeline.php line 87:
flarum     |                
flarum     |   Step failed  
flarum     |                
flarum     | 
flarum     | In PDOConnection.php line 31:
flarum     |                                                                                
flarum     |   SQLSTATE[HY000] [1045] Access denied for user 'flarum'@'172.24.0.5' (using   
flarum     |   password: YES)                                                               
flarum     |                                                                                
flarum     | 
flarum     | In PDOConnection.php line 27:
flarum     |                                                                                
flarum     |   SQLSTATE[HY000] [1045] Access denied for user 'flarum'@'172.24.0.5' (using   
flarum     |   password: YES)                                                               
flarum     |                                                                                
flarum     | 
flarum     | install [-f|--file FILE] [-c|--config CONFIG]
flarum     | 
flarum     | [INFO] End of flarum installation

```

**Comment by Hydrog3n** [comment]
Hello, check your password on flarum.env 
You set DB_PASS=testing but on composer file you set MYSQL_PASSWORD=testing123!

**Comment by TrevorKS** [comment]
> Hello, check your password on flarum.env
> You set DB_PASS=testing but on composer file you set MYSQL_PASSWORD=testing123!

I thought this was the first thing i checked -.-  I must have overlooked it. Good eye!

**Comment by TrevorKS** [comment]
Its still reporting some issues, but this can probably be closed. This particular issue has been fixed. Ill have to work to see whats going on tomorrow. Thanks all!
